capstone_project/src/main/java/com/google/sps/servlets/MapServlet.java

```diff
@@ -30,50 +38,66 @@
-                "https://news.uci.edu/2020/06/18/uci-podcast-jessica-millward-on-the-meaning-and-importance-of-juneteenth/"));
-    School uwb = new School("UWB", 47.758821, -122.190725, uwbNews);
-    schools.add(uwb);
-  }
+    for (Entity entity : results.asIterable()) {
+      String schoolName = (String) entity.getProperty("name");
+      double schoolLatitude = (double) entity.getProperty("latitude");
+      double schoolLongitude = (double) entity.getProperty("longitude");
+      School retrievedSchool = new School(schoolName, schoolLatitude, schoolLongitude);
 
-  /*Add some hard-coded School objects to the array of schools that will later be converted into markers on the map.*/
-  public MapServlet() {
-    addUCIToSchools();
-    addUWToSchools();
-    addUWBToSchools();
-  }
+      // Add the School loaded from Datastore to the schools list if it's not already in it.
+      if (!schools.contains(retrievedSchool)) {
+        schools.add(retrievedSchool);
+      }
+    }
 
-  @Override
-  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    // Send the School objects retrieved from Datastore.
     response.setContentType("application/json;");
     Gson gson = new Gson();
     response.getWriter().println(gson.toJson(schools));
   }
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+
+    // Initialize datastore object.
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+
+    // Get the input from the school entry form.
+    String userName = getParameter(request, "name-input", "");
+    double userLatitude = Double.parseDouble(getParameter(request, "latitude-input", ""));
+    double userLongitude = Double.parseDouble(getParameter(request, "longitude-input", ""));
+
+    // Check if the School is already in Datastore.
+    Filter nameFilter = new FilterPredicate("name", FilterOperator.EQUAL, userName);
+    Query query = new Query("School").setFilter(nameFilter);
+    PreparedQuery results = datastore.prepare(query);
+
+    // If it is not, create a schoolEntity for the School and add it to Datastore.
+    if (results.countEntities(FetchOptions.Builder.withDefaults()) == 0) {
+      Entity schoolEntity = new Entity("School");
+      schoolEntity.setProperty("name", userName);
+      schoolEntity.setProperty("latitude", userLatitude);
+      schoolEntity.setProperty("longitude", userLongitude);
+      datastore.put(schoolEntity);
+    }
+
+    // Respond with the result.
+    response.sendRedirect("pages/maps.html");
+  }
+
+  /** @return the request parameter, or the default value if not specified */
+  private String getParameter(HttpServletRequest request, String name, String defaultValue) {
+    String value = request.getParameter(name);
+    if (value == null) {
+      return defaultValue;
+    }
+    return value;
```

Comment:
nit: modify this to use the ternary operator `return value == null ? defaultValue : value;`
Also do you ever pass anything besides an empty string for defaultValue? If not, consider just making it a local variable.